### ISSUE TITLE
ACME V2 Help!

### ISSUE THREAD

**Issue by JNajera** [dokku/dokku-letsencrypt#183]
Hello, I run today on a problem when trying to update an SSL from a current website.

What should I do? Any help will be appreciated!

```
root@ubuntu-s-1vcpu-1gb-nyc1-01:~# dokku letsencrypt hidrotec_web
=====> Let's Encrypt hidrotec_web
-----> Updating letsencrypt docker image...
latest: Pulling from dokkupaas/letsencrypt-simp_le
Digest: sha256:95681f7cd659f23f451738121df9efe42ffc919e93a969781c40e936258fea72
Status: Image is up to date for dokkupaas/letsencrypt-simp_le:latest
       done updating
-----> Enabling ACME proxy for hidrotec_web...
-----> Getting letsencrypt certificate for hidrotec_web...
        - Domain 'hidrotec-web.apps-2.bolteam.com'
        - Domain 'www.hidrotec.com.mx'
        - Domain 'hidrotec.com.mx'
darkhttpd/1.12, copyright (c) 2003-2016 Emil Mikulic.
listening on: http://0.0.0.0:80/
2019-10-10 21:57:44,467:INFO:__main__:1211: Generating new account key
ACME server returned an error: urn:acme:error:unauthorized :: The client lacks sufficient authorization :: Account creation on ACMEv1 is disabled. Please upgrade your ACME client to a version that supports ACMEv2 / RFC 8555. See https://community.letsencrypt.org/t/end-of-life-plan-for-acmev1/88430 for details.


Debugging tips: -v improves output verbosity. Help is available under --help.
-----> Certificate retrieval failed!
-----> Disabling ACME proxy for hidrotec_web...
       done
```

**Comment by nabati** [comment]
Duplicate of https://github.com/dokku/dokku-letsencrypt/issues/181 but yes, the plugin seems completely broken right now.

**Comment by Cellane** [comment]
For now, this seems to be a temporary shutdown. According to [this thread](https://community.letsencrypt.org/t/end-of-life-plan-for-acmev1/88430), new registrations are disabled since yesterday, until later today. The same thing will happen next week, from Wednesday until Friday.

And at the end of the month, the old ACMEv1 API will be shut down for good.

In other words, the end is near, but as a very temporary solution, waiting a few hours should work around the problem. Until, of course, it hits us again once all of our certificates require a renewal.

**Comment by Cellane** [comment]
Okay, so this is probably kind of crazy, but…

I tried looking into easiest way of fixing this – after some research, it seems that while the original `simp_le` client is dead, there is a fork that supports ACMEv2 [available here](/zenhack/simp_le).

Before I waste your time, no, I was not (yet?) able to adjust this plugin to take advantage of the forked version. But I was able to perform all sorts of weird things to get a new certificate for my new subdomain.

Basically, since the ACMEv1 API is only closed for new registrations but will continue issuing certificate renewals until June of 2021, I started wondering if I can register my subdomain using the forked & updated version of `simp_le` and then simply transfer the generated `account_key.json` into my Dokku server and have it work that way. Surprise surprise, it actually works fine.

I’ll describe the commands I used below, rather briefly, since I don’t expect anyone to follow in my crazy footsteps. If you do and get stuck at some point, ping me and I’ll try to explain in greater detail!

```shell
## On your local machine
# ⬇️ This will probably not work for you due to my SSH configuration; simply clone the zenhack/simp_le repo 
$ git clone cellane.github.com:zenhack/simp_le
$ cd simp_le
$ docker image build -t cellane/simp_le:latest -f docker/Dockerfile .
$ docker container run -it cellane/simp_le sh

## In the built container
# ⬇️ This command will fail and that’s fine; it’ll still generate exactly what we want
$ simp_le --email <your@email.address> -f account_key.json -f account_reg.json -f fullchain.pem -f key.pem -d <your.domain.or.subdomain>:/var/www/html
# ⬇️ This should show the existance of `account_key.json` and `account_reg.json`
$ ls -l
# Do NOT exit the container just yet, rather open a new terminal tab/tmux session and continue

## In another shell
# Find container name, in my case it was `wizardly_aryabhata`, it’s the last column
$ docker container ls | grep simp_le
$ docker container cp wizardly_aryabhata:/simp_le/certs/account_key.json .
# ⬇️ Copy the file extracted from the Docker container to your Dokku server
$ scp account_key.json web:~

## On your Dokku server
# Here begins ultimate hackery…
$ sudo nvim /var/lib/dokku/plugins/available/letsencrypt/subcommands/default
# … go to line 92 and insert `sleep 20` in between the lines of `set +e` and `docker run --rm \`, you’ll understand in a moment
# ⬇️ Run this, but as soon as the domain name appears and the added `sleep` command gets hit, suspend the script with Ctrl-Z
$ dokku letsencrypt forum
$ cd /home/dokku/forum/letsencrypt/certs
# ⬇️ Verify that one folder with a random name was created, in my case let’s call it `asdf`
$ ls -l
$ sudo cp ~/account_key.json asdf/
# ⬇️ Resume certificate renewal (🤣) and after ~20 seconds, you should be good to go until 2021
$ fg
```

**Comment by FabienMotte** [comment]
Thank you so much @Cellane, worked for me too! 🙌

**Comment by Cellane** [comment]
@FabienMotte Woo! 🎉 It’s a total hackery, but I’m really glad it helped!

**Comment by dxpsite** [comment]
> # ⬇️ Copy the file extracted from the Docker container to your Dokku server
> $ scp account_key.json web:~

Ive got something troubles:

>root@subty:~# scp account_key.json web:~
>The authenticity of host 'web (93.180.10.106)' can't be established.
>RSA key fingerprint is SHA256:VkjPel2+Uee/9Tt4xJmHptp+qPZOi65YoGivclhDteE.
>Are you sure you want to continue connecting (yes/no)? yes
>Warning: Permanently added 'web,93.180.10.106' (RSA) to the list of known hosts.
>root@web's password:

**Comment by Cellane** [comment]
@dxpsite Ah, sorry! `web` is my SSH alias for the server that's running Dokku. Please replace `web` with whichever alias or host name you normally use!

**Comment by dxpsite** [comment]
MAN! YOU ARE GREAT! THANK YOU VERY MUCH, SAVED MY ASS !

**Comment by dean1012** [comment]
I am working on a more appropriate fix here: https://github.com/dokku/docker-letsencrypt-simp_le/pull/3

There's a few things at play here.

The Dockerfile did mention the zenhack fork which does support ACMEv2 but the startme script entry point by default overrides the zenhack version with the regular version which does not support ACMEv2. Seems a simple oversight.

Additionally, I had issues with the old `wget` method of obtaining the software so I replaced that method with a `git clone`  of zenhack in Dockerfile and a simple `git pull` if needed in the startme script.

I then built the Docker container and temporarily changed docker-letsencrypt to not pull the latest Docker image (so my testing version could be used). I verified the testing docker image had the correct zenhack version on it for ACMEv2.

I am running into a problem though that I'm not sure if it's in the docker image or this plugin. still looking:

```
=====> Let's Encrypt myapp
-----> Updating letsencrypt docker image...
       done updating
-----> Enabling ACME proxy for myapp...
-----> Getting letsencrypt certificate for myapp...
        - Domain 'myapp.example.com'
darkhttpd/1.12, copyright (c) 2003-2016 Emil Mikulic.
listening on: http://0.0.0.0:80/
Already up to date.
Selected IO plugins do not cover the following components: account_reg.

Debugging tips: -v improves output verbosity. Help is available under --help.
-----> Certificate retrieval failed!
-----> Disabling ACME proxy for myapp...
       done
```

**Comment by Cellane** [comment]
@dean1012 When I was trying to write a proper fix, I also ran into the same output you posted. I solved it by passing the `-f account reg.json` argument to the container, if I remember correctly.

**Comment by dean1012** [comment]
@Cellane yes I figured that out thanks!

My present problem is that it seems the container isn't saving the certs (or it is, but to the wrong location?)

Ultimately it says the cert is generated and says it is saving everything but only a `config` file exists so installation fails.

Running the underlying command with `-v` shows that certs are in fact being generated so the Let's Encrypt part is working.

The docker image, best I can tell, is fine how it is now with my pull request but I will need to submit another pull request to this repo (which i'll do as soon as I have it working) to utilize it correctly.

**Comment by dean1012** [comment]
@Cellane any thoughts on this?

```
=====> Let's Encrypt myapp                                                                                                                                                                                
-----> Updating letsencrypt docker image...                                                                                                                                                                          
       done updating                                                                                                                                                                                                 
-----> Enabling ACME proxy for myapp...
-----> Getting letsencrypt certificate for myapp...
        - Domain 'myapp.example.com'
-----> /home/dokku/myapp/letsencrypt/certs/bb9ed0ac322186d70b09a34091ce904529b245ad
darkhttpd/1.12, copyright (c) 2003-2016 Emil Mikulic.
listening on: http://0.0.0.0:80/
Already up to date.
2019-10-12 02:05:13,305:INFO:__main__:1323: Generating new account key
2019-10-12 02:05:14,403:INFO:__main__:1353: By using simp_le, you implicitly agree to the CA's terms of service: https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf                                  
2019-10-12 02:05:14,730:INFO:__main__:1414: Generating new certificate private key
2019-10-12 02:05:18,927:INFO:__main__:396: Saving account_key.json
2019-10-12 02:05:18,929:INFO:__main__:396: Saving account_reg.json
2019-10-12 02:05:18,929:INFO:__main__:396: Saving fullchain.pem
2019-10-12 02:05:18,930:INFO:__main__:396: Saving chain.pem
2019-10-12 02:05:18,930:INFO:__main__:396: Saving cert.pem
2019-10-12 02:05:18,931:INFO:__main__:396: Saving key.pem
-----> Certificate retrieved successfully.
-----> Installing let's encrypt certificates
-----> /home/dokku/myapp/letsencrypt/certs/bb9ed0ac322186d70b09a34091ce904529b245ad
CRT file specified not found, please check file paths
-----> Configuring myapp.example.com...(using built-in template)
-----> Creating http nginx.conf
-----> Running nginx-pre-reload
       Reloading nginx
-----> Disabling ACME proxy for myapp...
       done
```

The `/home/dokku/myapp/letsencrypt/certs/bb9ed0ac322186d70b09a34091ce904529b245ad` path referenced in the above output twice is from extra debugging statements I've put in. The first reference indicates the path that is passed to docker mounted to /certs within the container.

The second reference of the above path indicates the directory it is expecting to find the certs in for the symlink.

The contents of this directory:

```
user@host:~# ls /home/dokku/myapp/letsencrypt/certs/bb9ed0ac322186d70b09a34091ce904529b245ad
config
```

The contents of the `config` file:

```
--server https://acme-staging-v02.api.letsencrypt.org/directory --email me@example.com  -d myapp.example.com
```

**Comment by Cellane** [comment]
Hmm… At this moment, I don’t have much of an idea (nor do I think I’ll have much time this weekend, unfortunately), but seeing as it’s failing on finding a CRT file… and LE returns PEM files (`/home/dokku/<app>/letsencrypt/certs/current` should contain a lot of `pem` files)… but Nginx would surely want a CRT file… and `/home/dokku/<app>/tls` should contain `server.crt`…

I think maybe the certification is working perfectly fine (great job! 🎉), but the conversion from `pem` to `crt` fails for some reason, or it succeeds but puts the `crt` file in wrong location.

Sorry, I can’t be of more help at this moment.

**Comment by dean1012** [comment]
@Cellane I figured it out.

**Comment by dean1012** [comment]
I need @josegonzalez to merge these two pull requests:

* https://github.com/dokku/docker-letsencrypt-simp_le/pull/3
* https://github.com/dokku/dokku-letsencrypt/pull/184

After this is done, he will need to rebuild the Docker image located here:

* https://hub.docker.com/r/dokkupaas/letsencrypt-simp_le

After the above is all complete, updating the plugin on your system should cause the docker image to be pulled again and it should work with Acme V2 from then on out.

In the meantime @Cellane's work can be a workaround.

**Comment by Cellane** [comment]
Incredible job! 🎉

**Comment by josegonzalez** [comment]
Oh I should def move that image to the dokku namespace too...

**Comment by dean1012** [comment]
@josegonzalez If you plan to do that, I can add the docker pull change to my pull request if you'd like. Just let me know.

**Comment by dean1012** [comment]
As noted elsewhere, I went ahead and changed everything up to the dokku namespace so you'll have to rebuild the Docker image there.

**Comment by josegonzalez** [comment]
Closed by the work from @dean1012 and released as version `0.9.0` of the plugin.

**Comment by rsmithlal** [comment]
Thank you for the update!
